### PR TITLE
Ansible Mattermost Module

### LINKED ISSUE #21235: Support mattermost notification
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
module: mattermost (like slack)

##### ANSIBLE VERSION

```
ansible 2.2.1.0
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY

I would love to have a module to notify our mattermost installation (like http://docs.ansible.com/ansible/slack_module.html)

##### STEPS TO REPRODUCE

```yaml

```

##### ACTUAL RESULTS

```

```

### PR DESCRIPTION
**Pull request by bjolivot**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

mattermost

##### ANSIBLE VERSION

```
ansible 2.3.0 (mattermost 1a0011908c) last updated 2017/02/10 16:03:08 (GMT +200)
  config file = /home/ben/.ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY

New module for sending notifications to Mattermost.

This fixes #21235

### CODE REVIEW

**Comment by mikaoelitiana** [comment]
Thanks @bjolivot I am looking forward to seeing this merged

File: lib/ansible/modules/notification/mattermost.py
Comment (dagwieers) [inline lib/ansible/modules/notification/mattermost.py]: IMO it is probably better to use C(mattermost) instead, because the whole point of using M(something) is to have a link to the module's documentation. But that's where people already are when seeing the link. So a link pointing to itself is pretty moot.

Comment (dagwieers) [reply lib/ansible/modules/notification/mattermost.py]: That said, it seems some other modules do this too. Still a bad practice IMO.

File: lib/ansible/modules/notification/mattermost.py
Comment (dagwieers) [inline lib/ansible/modules/notification/mattermost.py]: I would drop the `required: false`. It is no longer needed to specify this for parameters that are not required.

File: lib/ansible/modules/notification/mattermost.py
Comment (dagwieers) [inline lib/ansible/modules/notification/mattermost.py:117]: Can we add check-mode support ?

Check-mode support in this case would probably do as much as possible without actually sending out the actual message. If there is a way to test sending as well (like some sort of /dev/null bucket) that would be great. But if not, if the module would test credentials and connection in check-mode, that in itself would be sufficient.

File: lib/ansible/modules/notification/mattermost.py
Comment (dagwieers) [inline lib/ansible/modules/notification/mattermost.py]: Nowadays, we list the imported functionality at the top of the module, and we explicitly list all the functionality we import. So in your case it would likely become:
```python
from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.urls import fetch_url
```

File: lib/ansible/modules/notification/mattermost.py
Comment (dagwieers) [inline lib/ansible/modules/notification/mattermost.py]: Most modules use the `result` variable for storing the result information for fail_json() or exit_json().
Only one module is using `rekwargs`: netconf_config. 262 using `result`, 27 using `ret`.

So up to you, but it's easier if people would use the same standard IMO.

File: lib/ansible/modules/notification/mattermost.py
Comment (dagwieers) [inline lib/ansible/modules/notification/mattermost.py]: Here as well. `required=False` is not strictly necessary. Sometimes it is easier to find what is required, if only those options have required :)

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by dagwieers** [comment]
@bjolivot If you add `This fixes #21235` into the description of the PR. The issue #21235 will automatically be closed.

File: lib/ansible/modules/notification/mattermost.py
Comment (bjolivot) [reply lib/ansible/modules/notification/mattermost.py]: done

File: lib/ansible/modules/notification/mattermost.py
Comment (bjolivot) [reply lib/ansible/modules/notification/mattermost.py]: done

File: lib/ansible/modules/notification/mattermost.py
Comment (bjolivot) [reply lib/ansible/modules/notification/mattermost.py]: done

File: lib/ansible/modules/notification/mattermost.py
Comment (bjolivot) [reply lib/ansible/modules/notification/mattermost.py:117]: Done
Message is not sent in case of check mode.

File: lib/ansible/modules/notification/mattermost.py
Comment (bjolivot) [reply lib/ansible/modules/notification/mattermost.py]: Done

File: lib/ansible/modules/notification/mattermost.py
Comment (bjolivot) [reply lib/ansible/modules/notification/mattermost.py]: Done
not sure if top of the module if before or after documentation
Moved before documentation.

**Comment by bjolivot** [comment]
ERROR: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN/ANSIBLE_METADATA. line 26

Ok, bad guess ^^

**Comment by dagwieers** [comment]
CI fails because the location of the imports apparently is strictly defined. Sigh :-)
So my remark "at the top" should be taken lightly ;-)
Sorry for that. Now that I know this too I will make it more clear in my next review.

File: lib/ansible/modules/notification/mattermost.py
Comment (dagwieers) [inline lib/ansible/modules/notification/mattermost.py]: Right, but now check_mode is not testing anything at all. We might as well have no check_mode in this code. I was hoping to see an implementation that looks like this: https://github.com/ansible/ansible/pull/19348/files#diff-473c3ae27a58aa5098bfcf29652164d8R141

So the solution implemented there was doing only the authentication to check connectivity and credentials (so it fails as if it would be running properly). Not sure if this is possible, but that would be as close as possible to the real thing.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/notification/mattermost.py
Comment (bjolivot) [reply lib/ansible/modules/notification/mattermost.py]: Webhook token validation is only made if text is provided & can't find /dev/null url/channel

What mattermost do:
 - if no text, 400 error, webhook token not verified
 - If text but bad webhook token, Http 500 error
 - if text given and good webhook, text is sent 


Don't know what to.
no check mode ?
add some tag ([check_mode])  in sent text ?
look for 500 error with forged token ?

Comment (dagwieers) [reply lib/ansible/modules/notification/mattermost.py]: Leave it as-is, but maybe add this information as comments ?
Maybe someone else can come around and fix it with your hints as a guide ?
Or maybe this is a valid feature-request to Mattermost.
A way to fake sending a message without sending the message.

Thanks for clarifying !

**Review by dagwieers — APPROVED** [review]

File: lib/ansible/modules/notification/mattermost.py
Comment (bjolivot) [reply lib/ansible/modules/notification/mattermost.py]: Note added :)

Comment (dagwieers) [reply lib/ansible/modules/notification/mattermost.py]: Ah, for me some inline comments, or a FIXME: statement was good enough.
I don't know if the end-user needs to deal with this TBH.

Beware also that the pylint check is expecting PEP8 so not too long lines :-/

File: lib/ansible/modules/notification/mattermost.py
Comment (gundalow) [inline lib/ansible/modules/notification/mattermost.py]: Sends notifications...

File: lib/ansible/modules/notification/mattermost.py
Comment (gundalow) [inline lib/ansible/modules/notification/mattermost.py]: All `description` lines (apart from short_description) should have full stops.

http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html

File: lib/ansible/modules/notification/mattermost.py
Comment (gundalow) [inline lib/ansible/modules/notification/mattermost.py]: No need to document defaults in the `description:` they will be pulled into the docs from the `default:` line

File: lib/ansible/modules/notification/mattermost.py
Comment (gundalow) [inline lib/ansible/modules/notification/mattermost.py]: No need to have the `required: false` lines, it's implicit.

File: lib/ansible/modules/notification/mattermost.py
Comment (gundalow) [inline lib/ansible/modules/notification/mattermost.py]: Document as `default:` and remove from `description:`

File: lib/ansible/modules/notification/mattermost.py
Comment (gundalow) [inline lib/ansible/modules/notification/mattermost.py]: You can remove this line

**Review by gundalow — CHANGES_REQUESTED** [review]
We've recently updated the "how to document your module" guide, might be worth having a look at that http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html

**Comment by bjolivot** [comment]
@gundalow thanks for your review
I've made update as you requested

**Comment by dagwieers** [comment]
shipit

File: lib/ansible/modules/notification/mattermost.py
Comment (gundalow) [inline lib/ansible/modules/notification/mattermost.py]: No need to repeat the module name:
`Sends notifications to U(http://your.mattermost.url) via the Incoming WebHook integration`

**Review by dagwieers — APPROVED** [review]

**Review by gundalow — APPROVED** [review]

**Comment by bjolivot** [comment]
What's happening now ? What do I need to do ?

**Comment by dagwieers** [comment]
Normally you need 2 shipits before this gets merged, but I am not sure mine count. You can use bot_status to see how many counts the bot has registered.

shipit

**Comment by gundalow** [comment]
shipit

File: lib/ansible/modules/notification/mattermost.py
Comment (dagwieers) [reply lib/ansible/modules/notification/mattermost.py:117]: Ok, not sure why this one is still open.

**Review by dagwieers — APPROVED** [review]

**Comment by dagwieers** [comment]
shipit

**Comment by dagwieers** [comment]
bot_status

**Comment by dagwieers** [comment]
@bjolivot I think the bot is waiting for you to say the magic word: ready_for_review

**Comment by bjolivot** [comment]
ready_for_review

**Comment by bjolivot** [comment]
bot_status

**Comment by bjolivot** [comment]
Could some maintainer send bot_status command ? 
I don't know if bot is still waiting from me...and what it is waiting for ^^

**Comment by dagwieers** [comment]
Maybe the bot is down ?
bot_status

**Comment by dagwieers** [comment]
If I look in the "Files changed" section, I still see one comment open. And I have no idea how to close that :-(

File: lib/ansible/modules/notification/mattermost.py
Comment (dagwieers) [reply lib/ansible/modules/notification/mattermost.py:117]: Test.

**Review by dagwieers — APPROVED** [review]

**Comment by bjolivot** [comment]
shipit

**Comment by dagwieers** [comment]
bot_broken

**Review by dagwieers — APPROVED** [review]
I approve everything. Pretty please.

**Comment by bjolivot** [comment]
the bot say he is waiting on me but i don't know why.

File: lib/ansible/modules/notification/mattermost.py
Comment (bjolivot) [reply lib/ansible/modules/notification/mattermost.py:117]: Test

**Comment by dagwieers** [comment]
Yeah, well I was assuming that maybe there was something from the review open.
And that that was why the bot was waiting for you to take action.

**Comment by gundalow** [comment]
Merged into `devel`, so this will be part of the Ansible 2.3 release

`MAINTAINERS.txt` updated: https://github.com/ansible/ansibullbot/pull/365
`CHANGELOG.md` updated: https://github.com/ansible/ansible/pull/21398

Thanks for the new module and reviewers

**Pull request merged**
